[Oceankoh] iP

Oceankoh opened the pull request:
# Revir
A fully integrated application to store manage tasks

## Features
- Track 3 different kinds of tasks
- Mark tasks as done/not done
- Automated persistent storage - Keep your data even after closing the application

## Usage
### Installation 
Download Revir [here](https://github.com/oceankoh/ip/releases/latest)
### Commands
- Adding tasks
```
todo <taskName>
deadline <taskName> /by <d/M/yyyy HHmm>
event <taskName> /from <d/M/yyyy HHmm> /to <d/M/yyyy HHmm>
```
- To show the task list
```
list
```
- To mark a task as done/not done:
```
mark <taskNo>
unmark <taskNo>
```
- To exit
```
bye
```
Features to add
- [ ] Create UI using Java FX
- [ ] Improve UX

## Additional Remarks
This is the current main method

```java
public class Revir {
    static TaskList taskList;
    static Storage storage;
    static Ui ui;
    static Parser parser;

    public static void main(String[] args) {
        parser = new Parser();
        ui = new Ui("Revir");
        storage = new Storage(Path.of("data", "tasks.dat"));
        taskList = new TaskList(storage, ui);
        ui.showWelcome();
        while (ui.isOpen()) {
            String input = ui.readInput();
            try {
                Command c = parser.parse(input);
                c.execute(ui, taskList);

            } catch (NumberFormatException e) {
                ui.showError("Invalid task index. Expected a number.");
            } catch (IOException e) {
                ui.showError("Unable to save file: " + e.getMessage());
            } catch (Exception e) {
                ui.showError(e.getMessage());
            }
        }
        ui.showExit();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Oceankoh committed:
renamed chatbot

Oceankoh committed:
echo user input

Oceankoh committed:
enabled user input history

Oceankoh committed:
mark and unmark tasks

Oceankoh committed:
add event types

Oceankoh committed:
generate test cases

Oceankoh committed:
use standard in instead of console

Oceankoh committed:
made print errors into exceptions

Oceankoh committed:
updated test cases

Oceankoh committed:
added support for delete

Oceankoh committed:
add delete test case

Oceankoh committed:
formatting

Oceankoh committed:
AI declaration

Oceankoh committed:
abstracted tasklist

Oceankoh committed:
save to file; adjusted test script to remove saves after each test

Oceankoh committed:
Merge branch 'branch-level-7'

Oceankoh committed:
Use dates and times instead of strings

Oceankoh committed:
Format dates and modify testcases

Oceankoh committed:
Merge branch 'branch-Level-8'

Oceankoh committed:
Abstract tasks, tasklist and storage

Oceankoh committed:
Abstract UI

Oceankoh committed:
Ignore .dat files

Oceankoh committed:
Abstract Parser; Add Command classes

Oceankoh committed:
Move everything into Revir package

Oceankoh committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Oceankoh committed:
Follow java package naming conventions

Oceankoh committed:
Update package name in test script

Oceankoh committed:
Add back java directory

Oceankoh committed:
Configure gradle

Oceankoh committed:
Add JavaDoc for storage class

Oceankoh committed:
Add JavaDoc for Exceptions

Oceankoh committed:
Add JavaDoc for task package

Oceankoh committed:
Add JavaDoc for command package

Oceankoh committed:
Update AI usage observations

Oceankoh committed:
Format code

Oceankoh committed:
Implement find command

Oceankoh committed:
Merge branch 'branch-A-JavaDoc'

Oceankoh committed:
Merge branch 'branch-A-CodingStandard'

Oceankoh committed:
Merge branch 'branch-Level-9'

Oceankoh committed:
Add JavaDoc for find method and format

Oceankoh committed:
Add JUnit tests for find, create and delete

Oceankoh committed:
Checkstyle config

Oceankoh committed:
Adhere to checkstyle

yockcheng commented on src/main/java/revir/Revir.java:
```diff
@@ -0,0 +1,39 @@
+        ui.showExit();
+    }
+
+}
```
I like the naming of your methods, it is easy to understand what the method is expected to do.

yockcheng commented on src/main/java/revir/system/Storage.java:
```diff
@@ -0,0 +1,58 @@
+import java.nio.file.Path;
+import revir.tasks.Task;
+import java.util.ArrayList;
+
```
Perhaps you could order your imports in lexicographical order?

yockcheng commented on src/main/java/revir/system/Storage.java:
```diff
@@ -0,0 +1,58 @@
+        if (Files.exists(this.path)) {
+            ObjectInputStream ois = new ObjectInputStream(new FileInputStream(this.path.toString()));
+            ArrayList<Task> result = (ArrayList<Task>) ois.readObject();
+            ois.close();
```
Perhaps you could consider a more specific variable name instead of result?

yockcheng commented on src/main/java/revir/user/Parser.java:
```diff
@@ -0,0 +1,127 @@
+        Action action = Parser.actionMap.getOrDefault(commandWord, Action.INVALID);
+        // use switch case to return different methods based on command type
+        switch (action) {
+            case BYE:
```
Should the indentation of the "cases" be in line with "switch"?

yockcheng commented on src/main/java/revir/user/command/Create.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Represents a command to create a new task and add it to the task list.
+ */
+public class Create extends Command {
```
Perhaps the class names for the commands could be more specific?

yockcheng commented on src/test/java/revir/user/command/CreateCommandTest.java:
```diff
@@ -0,0 +1,65 @@
+
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
I like the way you included empty lines to separate and structure the import statements which made it a lot clearer.

yockcheng reviewed (commented):
Overall, I really liked the names you chose for your methods and variables. It was easy to understand what the function is likely to be doing before even looking at the implementation of the code.

Oceankoh committed:
Fix minor UI bugs

Oceankoh committed:
Format code according to checkstyle guidelines

Oceankoh committed:
Adjust testcase to new output format

Oceankoh committed:
Merge branch 'checkstyle'

myrilla12 commented on src/main/java/revir/Revir.java:
```diff
@@ -0,0 +1,46 @@
+import revir.user.command.Command;
+
+/**
+ * The Revir class is the main class that runs the Revir program.
```
Good detailed header comment but try to start header comments with something like 'Initializes' in compliance with JavaDoc comments structure. I noticed this for the other class header comments as well.

myrilla12 commented on src/main/java/revir/system/Storage.java:
```diff
@@ -0,0 +1,62 @@
+    /**
+     * Saves the given list of tasks to a file.
+     *
+     * @param tasks the list of tasks to be saved
```
Some JavaDoc comments for parameter and return are missing a full stop at the end.

myrilla12 commented on src/main/java/revir/user/Parser.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    private static DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+    private static final Map<String, Action> actionMap = new HashMap<>();
```
Should final variables be named in SCREAMING_SNAKE_CASE according to Java conventions?

myrilla12 commented on src/main/java/revir/user/Parser.java:
```diff
@@ -0,0 +1,144 @@
+     * @throws InvalidFormatException If the command is recognized but the format is invalid.
+     */
+    public Command parse(String commandString) throws IllegalCommandException, InvalidFormatException {
+        String[] split = commandString.trim().split(" ", 2);
```
'split' seems like a suitable name to me, but according to Java naming conventions, collections should be named in plural.

myrilla12 commented on src/main/java/revir/user/Parser.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    Command parseDeadline(String args) throws InvalidFormatException {
+        String[] taskInfo = args.split(" /by ");
```
Same over here with the plural naming

myrilla12 reviewed (commented):
Overall the code looks nice and clean to me. I just spot some minor things that don't follow convention in the JavaDoc comments and some naming.

Oceankoh committed:
Add Gui support

Oceankoh committed:
Merge branch 'branch-Level-10'

Oceankoh committed:
Fix faulty images

Oceankoh committed:
Add assertions

Assert existence of Gui window before calling sendMessage
Assert existence of Gui instance to handle user input from MainWindow
Assert unreachable in command parser for exhausive command checks

Oceankoh committed:
Remove unnecessary method

Oceankoh committed:
Rename variable to abide by code quality guidelines

Oceankoh committed:
Refactor code for better code quality

Oceankoh committed:
Concat print statements

Oceankoh committed:
Prevent adding of duplicate tasks

Oceankoh committed:
Merge pull request #2 from Oceankoh/branch-A-Assertions

Assert existence of Gui window before calling sendMessage
Assert existence of Gui instance to handle user input from MainWindow
Assert unreachable in command parser for exhaustive command checks

Oceankoh committed:
Merge branch 'master' into branch-A-CodeQuality

Oceankoh committed:
Merge branch 'master' into branch-C-DetectDuplicates

Oceankoh committed:
Merge pull request #3 from Oceankoh/branch-A-CodeQuality

Code quality improvements

Oceankoh committed:
Merge branch 'master' into branch-C-DetectDuplicates

Oceankoh committed:
Merge pull request #1 from Oceankoh/branch-C-DetectDuplicates

Prevent adding of duplicate tasks

Oceankoh committed:
Do not show error for corrupted tasklist

NullPointerException when loading a corrupted tasklist.

Gui has not been launched and error messsage cannot be shown.
There is a cyclic dependency of tasklist and ui requiring each other to
be initialized.

Ui instance is not needed if error message is not shown.

Oceankoh committed:
Create user guide

Oceankoh committed:
Fix task mark bug

Oceankoh committed:
Rename booleans to abide by naming conventions

Oceankoh committed:
Remove whitespaces according to checkstyle guidelines
